response.go

```diff
@@ -0,0 +1,255 @@
+//		x400Address                     [3]     ORAddress,
+//		directoryName                   [4]     Name,
+//		ediPartyName                    [5]     EDIPartyName,
+//		uniformResourceIdentifier       [6]     IA5String,
+//		iPAddress                       [7]     OCTET STRING,
+//		registeredID                    [8]     OBJECT IDENTIFIER }
+type generalNames struct {
+	Name asn1.RawValue `asn1:"optional,tag:4"`
+}
+
+// MarshalBinary encodes the response to binary form.
+// This method implements encoding.BinaryMarshaler.
+//
+// Reference: https://pkg.go.dev/encoding#BinaryMarshaler
+func (r *Response) MarshalBinary() ([]byte, error) {
+	if r == nil {
+		return nil, errors.New("nil response")
+	}
+	return asn1.Marshal(*r)
+}
+
+// UnmarshalBinary decodes the response from binary form.
+// This method implements encoding.BinaryUnmarshaler.
+//
+// Reference: https://pkg.go.dev/encoding#BinaryUnmarshaler
+func (r *Response) UnmarshalBinary(data []byte) error {
+	_, err := asn1.Unmarshal(data, r)
+	return err
+}
+
+// SignedToken returns the timestamp token with signatures.
+//
+// Callers should invoke SignedToken.Verify to verify the content before
+// comsumption.
+func (r *Response) SignedToken() (*SignedToken, error) {
+	if err := r.validateStatus(); err != nil {
+		return nil, err
+	}
+	return ParseSignedToken(r.TimeStampToken.FullBytes)
+}
+
+// Validate checks if resp is a successful timestamp response against
+// its corresponding request based on RFC 3161.
+// It is used when a timestamp requestor receives the response from TSA.
+func (resp *Response) Validate(req *Request) error {
+	if req == nil {
+		return &InvalidResponseError{Msg: "missing corresponding request"}
+	}
+	if resp == nil {
+		return &InvalidResponseError{Msg: "response cannot be nil"}
+	}
+	if err := resp.validateStatus(); err != nil {
+		return err
+	}
+	token, err := resp.SignedToken()
+	if err != nil {
+		return &InvalidResponseError{Detail: err}
+	}
+	info, err := token.Info()
+	if err != nil {
+		return &InvalidResponseError{Detail: err}
+	}
+	if info.Version != 1 {
+		return &InvalidResponseError{Msg: fmt.Sprintf("timestamp token info version must be 1, but got %d", info.Version)}
+	}
+	// check policy
+	if req.ReqPolicy != nil && !req.ReqPolicy.Equal(info.Policy) {
+		return &InvalidResponseError{Msg: fmt.Sprintf("policy in response %v does not match policy in request %v", info.Policy, req.ReqPolicy)}
+	}
+	// check MessageImprint
+	if !info.MessageImprint.HashAlgorithm.Algorithm.Equal(req.MessageImprint.HashAlgorithm.Algorithm) ||
+		!bytes.Equal(info.MessageImprint.HashedMessage, req.MessageImprint.HashedMessage) {
+		return &InvalidResponseError{Msg: fmt.Sprintf("message imprint in response %+v does not match with request %+v", info.MessageImprint, req.MessageImprint)}
+	}
+	// check gen time
+	genTime := info.GenTime
+	if genTime.Location() != time.UTC {
+		return &InvalidResponseError{Msg: "TSTInfo genTime must be in UTC"}
+	}
```

Comment:
What's the reference for this check? If it is not in UTC, can we just call `UTC()`?

Reply:
Reference is: https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2
```
genTime is the time at which the time-stamp token has been created by
   the TSA.  It is expressed as UTC time (Coordinated Universal Time) to
   reduce confusion with the local time zone use.
```